models/clan.go

```diff
@@ -978,10 +978,18 @@ func SearchClan(
 		return clans, nil
 	}
 
+	var filter interface{}
+	if useTextSearch {
+		filter = bson.M{"$text": bson.M{"$search": term}}
+	} else {
+		escapedTerm := fmt.Sprintf(`\Q%s\E`, term)
+		filter = bson.M{"name": bson.M{"$regex": escapedTerm, "$options": "i"}}
```

Comment:
This query is very inefficient, be careful. It was like this before I changed it to use text index!

Reply:
Here: https://docs.mongodb.com/manual/reference/operator/query/regex/#index-use

Reply:
I realize that, however the way we currently do it is by essentially by-passing khan to execute that search. Text index doesn't work as intended for our purposes.

Reply:
Would it be acceptable to change to case-sensitive then?

Reply:
> For case sensitive regular expression queries, if an index exists for the field, then MongoDB matches the regular expression against the values in the index, which can be faster than a collection scan.

As I understand, the search is still O(N) but with a cheaper constant, because you load index data instead of the documents themselves, which is supposed to be faster... I think only load tests would tell if it's really okay. But, as I remember, most queries would take more than 2 seconds with the regex of this patch...

Reply:
How about using a prefix expression? "^\Qterm\E"

Reply:
>  Further optimization can occur if the regular expression is a “prefix expression”, which means that all potential matches start with the same string. This allows MongoDB to construct a “range” from that prefix and only match against those values from the index that fall within that range.

Actually I think it's a good idea! As I understand, it's still O(N), but only in the amount of strings starting with the prefix you gave to the query, which is probably much much much less than the total amount of strings. But I don't have experience enough to ensure that, I'd still make some load tests to make sure we don't just trust in our interpretation of the docs, though. You'll take less than a day to make those load tests using this: https://github.com/topfreegames/khan/tree/master/loadtest

Reply:
Also I think it still needs do be case-sensitive.

Reply:
What are \Q and \E for? I'm not 100% sure they wouldn't prevent the prefix expression to work as we expect.

Reply:
Yes, I wasn't excluding case-sensitivity. Sorry if it wasn't clear.
\Q and \E are used to escape the term, in the sense of interpreting literally all the characters in between. https://www.pcre.org/current/doc/html/pcre2syntax.html#SEC2

Reply:
> Yes, I wasn't excluding case-sensitivity. Sorry if it wasn't clear.
\Q and \E are used to escape the term, in the sense of interpreting literally all the characters in between. https://www.pcre.org/current/doc/html/pcre2syntax.html#SEC2

Got it. Still can't tell if it will have an impact or not... But the MongoDB docs do not have any warnings on that, so I guess it's okay. Anyway, only load tests will really tell if a solution is good or not.

Reply:
I'm having trouble figuring out exactly what to do for those load tests. Am I supposed to write a new config file for the load test? How do I know what values would be reasonable? And after perfoming those load tests, how will I know if the result is acceptable or not?

Reply:
1. Setup a new instance of Khan API to load test it. New instances of Redis, Postgres and MongoDB will also be necessary for this API to work for your purposes, because clan insertions on Postgres will be dispatched to MongoDB through Redis.
2. Make sure this new API instance will send its latency metrics to somewhere (Datadog, Graphana...). You're interested in the latency metric of the Search Clans endpoint.
3. Drill this API somehow (from your local machine or from a cluster) using [the load tests command-line tool](https://github.com/topfreegames/khan/tree/master/loadtest) I created. For this step you just need to copy/change [the tool's configs](https://github.com/topfreegames/khan/blob/master/config/local.yaml#L56). What you want here is to create a lot of clans first, and then drill the Search Clans endpoint. For that you can first set the probability of `createClan` to 1 and everything else to 0, then later set the probability of `searchClans` to 1 and everything else to 0.
4. After drilling the Search Clans endpoint, analyze its latency metric along the experiment. You want to know if this metric did not reach absurd values, like 2 seconds. Also, check the databases' metrics (CPU, etc.) to check if they were okay.

Someone in your team can certainly help you with the details in each of these steps. Also, refer to [the docs of the load test command-line tool](https://github.com/topfreegames/khan/blob/master/loadtest/README.md), I'm pretty sure I explained all its configs there. If something is not clear, feel free to ask me!

Reply:
Thanks @matheuscscp. We'll first see if the load test is really required since it is already running on one of our games, its 95p latency seems to be  around 200-300ms.